### ISSUE TITLE
New data point for invalid UTF-8 sequence

### ISSUE THREAD

**Issue by JockLawrie** [JuliaDB/ODBC.jl#122]
Hi there,
I'm connecting from a CentOS VM to a remote SQL Server instance.
I'm using Julia 0.4.3 and the master branch of ODBC.jl.
The connection works fine, but I get an error when trying to retrieve the result - see below.
I understand there has been much talk about encodings in this context. Just thought this may provide an extra data point. Thoughts?

```
using ODBC
con = ODBC.connect("dsn_name", usr="username", pwd="mypwd")
ODBC Connection Object
----------------------
Connection Data Source: dsn_name
dsn_name Connection Number: 1
Contains resultset(s)? No

df = ODBC.query("select top 10 * from MyTable", con);
ERROR: UnicodeError: invalid UTF-8 sequence starting at index 80 (0x27 is not a continuation byte)
 in __unsafe_checkstring#18__ at ./unicode/checkstring.jl:78
 in __unsafe_checkstring#16__ at ./unicode/checkstring.jl:66
 in convert at ./unicode/utf8.jl:243
 in ODBCClean at /home/jock.lawrie/.julia/v0.4/ODBC/src/backend.jl:127
 in ODBCError at /home/jock.lawrie/.julia/v0.4/ODBC/src/backend.jl:244
 in ODBCQueryExecute at /home/jock.lawrie/.julia/v0.4/ODBC/src/backend.jl:44
 in query at /home/jock.lawrie/.julia/v0.4/ODBC/src/userfacing.jl:62
```

**Comment by ScottPJones** [comment]
The problem is simply that you the data you are reading is not valid UTF-8, which the conversion code is detecting.
Can you check with some other tool just what the contents of `MyTable` is, and what the character set being used is?  Ansi Latin 1?  CP1252? (those are the most common 8-bit character sets I've seen).

**Comment by StefanKarpinski** [comment]
The hacky solution for now is to just go into the ODBC source and change this line:

https://github.com/JuliaDB/ODBC.jl/blob/a36ef63519b98729a4c11/src/backend.jl#L43

So that it calls `utf8` instead of `utf16`. This is an issue with using the "ANSI" API for the SQL Server ODBC driver, which expects UTF-8 encoded data where other ODBC drivers expect UTF-16 data. Should be fixed by @quinnj's overhaul of the package to use newer APIs.

**Comment by ScottPJones** [comment]
I think that will only help if the SQL Server data is encoded as UTF-8 though.
If it is CP1252 or ISO/ANSI Latin 1, it will still get errors.
Hopefully Jacob can get the fix in soon to use the correct APIs.

**Comment by JockLawrie** [comment]
Not exactly sure what I'm doing but here's what I did using R. Hope it helps.

```
library(RODBC)
con <-odbcConnect("dsn_name", uid="username", pwd="mypwd")
df <- sqlQuery(con, "select top 10 * from MyTable")    # All good, no error, just checking.
df <- sqlQuery(con, "select distinct character_set_name, collation_name from information_schema.columns")
close(con)
```

The result of the 2nd query was:
```
> df
character_set_name    collation_name
    NA                  NA
    iso_1               SQL_Latin1_General_CP1_CI_AS
    UNICODE             SQL_Latin1_General_CP1_CI_AS
```

**Comment by ScottPJones** [comment]
Yep, as I suspected, you have data encoded as ANSI/ISO Latin 1.
Changing the code to have utf8 instead of utf16 won't help in that case.
Using the W API ensures that you get either UTF-16 or UTF-32, depending on what SQLWCHAR is for the ODBC client.

**Comment by StefanKarpinski** [comment]
@JockLawrie, can you try my patch and see if it helps?

**Comment by ScottPJones** [comment]
That might not cause an error, however, utf8 doesn't do ANSI to utf8 translations, so you'd just be getting garbage for characters > 0x7f.

**Comment by JockLawrie** [comment]
I re-ran the code in the first post of this issue, preceded by Pkg.update(). Same result I'm afraid.

**Comment by StefanKarpinski** [comment]
@ScottPJones: as the package stands, this error would happen for plain old ASCII data, but it's entirely possible that this table only has ASCII data in it, in which case the change I proposed would help.

@JockLawrie: I'm unclear on whether you changed the line I linked to or not before doing that.

**Comment by JockLawrie** [comment]
@StefanKarpinski: Success! Your patch worked, thanks.

Though I do get some funny formatting which is of course a separate issue (or user error). E.g., in R I get dates as yyyy-mm-dd HH:MM:SS, whereas in Julia I get ODBC.SQLTimestamp(yyyy,mm,dd,HH,MM,SS,ms).

**Comment by JockLawrie** [comment]
There is one column that isn't returned faithfully. It's not important for my purposes but may provide some insight. The column name is RowVersion.
In R the first entry in the column reads: 0000000000200dda.
In Julia it reads: "\0\0\0\0\0 \r?", including quotations, and the ? is not the ascii character.

**Comment by StefanKarpinski** [comment]
What are the types of the columns that are still giving issues? I.e. the timestamp one and RowVersion?

Another thing you could try is doing `Pkg.checkout("ODBC", "jq/remodel")` and seeing if @quinnj's new branch fixes more things. It's still being tested out by various people, but it might be better.

**Comment by JockLawrie** [comment]
The timestamp column has type "datetime".
The RowVersion column has type "timestamp".

I tried the jq/remodel branch, here's what happened:
```
> con = ODBC.DSN("dsn_name", "username", "mypwd")
ODBC.DSN(dsn_name)

> df = ODBC.query(con, "select top 10 * from MyTable")
ERROR: UndefVarError: ODBCQueryExecute not defined
  in query at /home/jock.lawrie/.julia/v0.4/ODBC/src/userfacing.jl:9
  in query at /home/jock.lawrie/.julia/v0.4/ODBC/src/userfacing.jl:8
```

**Comment by quinnj** [comment]
@JockLawrie, try this instead on the `jq/remodel` branch.

```julia
source = ODBC.Source(con, "select top 10 * from MyTable")
Data.stream!(source, Data.Table)
```

**Comment by JockLawrie** [comment]
@quinnj I tried your suggestion. The results are largely the same, except:
- The first entry in the column with type datetime is: -19088:18157:32636T:-2233618128
- Each entry in the "RowVersion" column (with type timestamp) is #NULL

Happy to help test further.
Also, any tips on viewing/printing a Data.Table?

**Comment by quinnj** [comment]
should be resolved by 7ae7ebfac60ed3ffb1ccefd552993a3073f8b77b
Please re-open if you see any issues

**Comment by ax1ine** [comment]
Reopening again :( Ubuntu x64, ODBC.jl with Firebird 3.
`sql = "SELECT * FROM CRUISE"`

I get this error:
`ERROR: LoadError: HY000: [ODBC Firebird Driver][Firebird]Dynamic SQL Error
SQL error code = -104
Unexpected end of command - line 1, column 17
Token unknown - line 1, column 10
F`

Which clearly shows that Firebird receives just 1/2 of the query (UTF8 required, UTF16 is used). I tried different queries but it's all the same - a half of the text is being passed to the server. 

Seems like the code has been reworked and all the old 'hacky' methods don't work any more. How and where do I specify that I need UTF8?
